geometric_kernels/spaces/so.py

```diff
@@ -0,0 +1,368 @@
+            rep_dim = int(reduce(operator.mul, (2 / math.factorial(2 * k) for k in range(1, self.rank)))
+                          * reduce(operator.mul, ((qs[i] - qs[j]) * (qs[i] + qs[j])
+                                                  for i, j in itertools.combinations(range(self.rank), 2)), 1))
+            return int(round(rep_dim))
+
+    def _compute_eigenvalue(self, signature):
+        np_sgn = np.array(signature)
+        rho = self.rho
+        eigenvalue = (np.linalg.norm(rho + np_sgn) ** 2 - np.linalg.norm(rho) ** 2)  # / killing_form_coeff
+        return eigenvalue.item()
+
+    def _torus_representative(self, X):
+            """The function maps Lie Group Element X to T -- a maximal torus of the Lie group
+            [n1,n2,\ldots, nk,X, X] ---> [n1,n2,\ldots,nk,X, X]"""
+            if self.n == 3:
+                # In SO(3) the torus representative is determined by the non-trivial pair of eigenvalues,
+                # which can be calculated from the trace
+                trace = einsum('...ii->...', X)
+                real = (trace - 1) / 2
+                imag = B.sqrt(B.max(1 - B.square(real), B.zeros_like(real)))
+                return B.view_as_complex(B.cat((real.unsqueeze(-1), imag.unsqueeze(-1)), -1)).unsqueeze(-1)
+            elif self.n % 2 == 1:
+                # In SO(2n+1) the torus representative is determined by the (unordered) non-trivial eigenvalues
+                eigvals = B.linalg.eigvals(X)
+                sorted_ind = B.sort(B.view_as_real(eigvals), dim=-2).indices[..., 0]
+                eigvals = B.gather(eigvals, dim=-1, index=sorted_ind)
+                gamma = eigvals[..., 0:-1:2]
+                return gamma
+            else:
+                # In SO(2n) each unordered set of eigenvalues determines two conjugacy classes
+                eigvals, eigvecs = B.linalg.eig(X)
+                sorted_ind = B.sort(B.view_as_real(eigvals), dim=-2).indices[..., 0]
+                eigvals = B.gather(eigvals, dim=-1, index=sorted_ind)
+                eigvecs = B.gather(eigvecs, dim=-1, index=sorted_ind.unsqueeze(-2).broadcast_to(eigvecs.shape))
+                # c is a matrix transforming x into its canonical form (with 2x2 blocks)
+                c = B.zeros_like(eigvecs)
+                c[..., ::2] = eigvecs[..., ::2] + eigvecs[..., 1::2]
+                c[..., 1::2] = (eigvecs[..., ::2] - eigvecs[..., 1::2])
+                # eigenvectors calculated by LAPACK are either real or purely imaginary, make everything real
+                # WARNING: might depend on the implementation of the eigendecomposition!
+                c = c.real + c.imag
+                # normalize s.t. det(c)≈±1, probably unnecessary
+                c /= math.sqrt(2)
+                B.pow(eigvals[..., 0], B.det(c).sgn(), out=eigvals[..., 0])
+                gamma = eigvals[..., ::2]
+                return gamma
+
+    def _compute_character_formula(self, signature):
```

Comment:
This seems to not belong here, is it needed anywhere except the script that precomputes the characters? If not, consider moving it there. If yes, consider moving it to `utils` or somewhere.